python/tank/bootstrap/manager.py

```diff
@@ -221,11 +244,50 @@ def set_progress_callback(self, callback):
         """
         self._progress_cb = callback
 
-    def bootstrap_engine(self, engine_name, entity=None):
+    def bootstrap_engine(self,
+                         engine_name,
+                         entity=None,
+                         progress_callback=None,
+                         completed_callback=None,
+                         failed_callback=None):
         """
         Create an sgtk instance for the given project or site,
         then launch into the given engine.
 
+        When property ``async_bootstrap`` is false, the whole engine bootstrap logic
+        will be executed synchronously in the main application thread. When true,
+        an :class:`sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+
+        When property ``async_bootstrap`` is true, 3 callback functions should be provided.
```

Comment:
This makes me think a separate method might be preferable. 

If we stick with this, I'm wondering if we could get away with a single callback with statuses. i.e.:

`bootstrap_engine(engine_name, entity, callback)` where `callback` has a signature like this:

`callback(status, step, total_steps, message=None)`

 Status could be exposed as `ToolkitManager.PROGRESS`, `ToolkitManager.COMPLETE`, and `ToolkitManager.FAILURE`. Maybe we could handle progress, completion, and failure with something like that? What do you think?